### PR TITLE
Button's setAlignment() is being considered and right margin value is recalculated depending on alignment set.

### PR DESCRIPTION
**Pull request by deepika-u**
Button's setAlignment() is calculating the right margin value now when text and image are set depending on the alignment of the button.

Fixes
https://github.com/eclipse-platform/eclipse.platform.swt/issues/519

@niraj-modi 
@sravanlakkimsetti 
Can anyone of you review the changes and let me know your comments please.

### CODE REVIEW

File: bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Button.java
Comment (niraj-modi) [inline bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Button.java:317]: - These comments are not useful should be removed.
- Also if-else can be simplified as below:
if ((style & SWT.LEFT) != 0) {
	margin = MARGIN;
} else if ((style & SWT.RIGHT) != 0) {
	margin = Math.max (MARGIN, (rect.right - rect.left - margin - MARGIN));
} else {
	margin = Math.max (MARGIN, (rect.right - rect.left - margin) / 2);
}

**Review by niraj-modi — CHANGES_REQUESTED** [review]

**Review by iloveeclipse — DISMISSED** [review]
According to the bug report, this is not only a Windows issue, so we also need a fix for Linux. Can't say anything about Mac, because I have no access to it.

**Review by niraj-modi — COMMENTED** [review]
@deepika-u 
Please verify if this fix works well on Windows 200% zoom as well ?

**Comment by niraj-modi** [comment]
> According to the bug report, this is not only a Windows issue, so we also need a fix for Linux. Can't say anything about Mac, because I have no access to it.

@elsazac 
Please check/confirm if you see this problem on MAC.

**Comment by elsazac** [comment]
@niraj-modi 

As Philipus said, On Mac, the image is on the left and the text is getting aligned when we set the alignment to left right and center.

**Comment by deepika-u** [comment]
> @deepika-u Please verify if this fix works well on Windows 200% zoom as well ?

I have tested the pr with 200% scaling level and the alignment works well even in this case on windows 11.

**Comment by deepika-u** [comment]
Hi @ptziegler

Can you also test this pr on windows and confirm if this fix works fine please? 

I shall go ahead opening platform specific issues for others to check on mac and linux independently as i have only windows environment right now.

Thanks alot.

**Comment by vogella** [comment]
@niraj-modi not sure if you know this but you can trigger a rebase for a commit without conflicts via the web UI: 
image

**Comment by deepika-u** [comment]
@iloveeclipse : Can you please review this windows pr and merge it please.

For Linux an issue is open => https://github.com/eclipse-platform/eclipse.platform.swt/issues/927
For Mac and issue is open => https://github.com/eclipse-platform/eclipse.platform.swt/issues/928

**Comment by deepika-u** [comment]
@iloveeclipse 
Hi Andrey, Can you find some time to review this pr and merge it please.

**Comment by iloveeclipse** [comment]
> @iloveeclipse Hi Andrey, Can you find some time to review this pr and merge it please.

Can you please post screenshot before / after the changes for some meaningful SWT snippet?

**Comment by deepika-u** [comment]
> > @iloveeclipse Hi Andrey, Can you find some time to review this pr and merge it please.
> 
> Can you please post screenshot before / after the changes for some meaningful SWT snippet?

I have tried the below snippet to recreate the problem locally

```
package org.eclipse.swt.bugs3;
import org.eclipse.swt.*;
import org.eclipse.swt.graphics.*;
import org.eclipse.swt.layout.*;
import org.eclipse.swt.widgets.*;

public class ButtonAlignmentExample {

	public static void main(String[] args) {
		Display display = new Display();
		Shell shell = new Shell(display);
		shell.setText("Button Alignment Example");
		shell.setLayout(new GridLayout(1, false));

		// Load an image (you should replace "imagePath" with your image file path)
		//Image image = ImageDescriptor.createFromFile(Snippet.class, "error_tsk.png").createImage();
		Image image = new Image(display, "C:\\Users\\02236C744\\Desktop\\error_tsk.png");

		// Create a button with text and image
		Button button = new Button(shell, SWT.PUSH);
		button.setText("Button Text");
		button.setImage(image);
		button.setSize(200, 50);

		// Set the alignment to CENTER
		button.setAlignment(SWT.RIGHT);
		//button.setAlignment(SWT.LEFT);
		//button.setAlignment(SWT.CENTER);

		// Set layout data to take up the whole shell
		button.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));

		shell.setSize(200, 50);
		shell.open();
		shell.pack();
		while (!shell.isDisposed()) {
			if (!display.readAndDispatch()) {
				display.sleep();
			}
		}

		// Dispose of the image and resources
		image.dispose();
		display.dispose();
	}
}
```

Before the fix, i am seeing like this
Before_Fix


After the fix, i am seeing like this
After_the fix


Please let me know if i was complete in answering you. Thanks.

**Comment by iloveeclipse** [comment]
> Please let me know if i was complete in answering you. Thanks.

May be I understand something wrong.
First, this PR is about Linux but you show me Windows screenshort.
Second, please add **improved** test that shows the change under  `org.eclipse.swt.tests/ManualTests/org/eclipse/swt/tests/manual`
Third, here is the slightly modified example I tried and it shows something unexpected. 
Please feel free to update the example to show all three buttons in one shell so one can clearly see the difference between them.

```
package org.eclipse.swt.tests.manual;
import org.eclipse.swt.SWT;
import org.eclipse.swt.graphics.Image;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Button;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Shell;

public class ButtonAlignmentExample {

	public static void main(String[] args) {
		Display display = new Display();
		Shell shell = new Shell(display);
		shell.setText("Button Alignment Example");
		shell.setLayout(new GridLayout(1, false));

		// Load an image (you should replace "imagePath" with your image file path)
		//Image image = ImageDescriptor.createFromFile(Snippet.class, "error_tsk.png").createImage();
		Image image = new Image(display, "data/eclipse32.png");

		// Create a button with text and image
		Button button;
		button = new Button(shell, SWT.PUSH);
		button.setText("Right Text");
		button.setImage(image);
		button.setSize(200, 50);
		button.setAlignment(SWT.RIGHT);
		button.setLayoutData(new GridData(SWT.FILL, SWT.FILL, false, false));

		button = new Button(shell, SWT.PUSH);
		button.setText("Left Text");
		button.setImage(image);
		button.setSize(200, 50);
		button.setAlignment(SWT.LEFT);
		button.setLayoutData(new GridData(SWT.FILL, SWT.FILL, false, false));

		button = new Button(shell, SWT.PUSH);
		button.setText("Center Text");
		button.setImage(image);
		button.setSize(200, 50);
		button.setAlignment(SWT.CENTER);

		// Set layout data to take up the whole shell
		button.setLayoutData(new GridData(SWT.FILL, SWT.FILL, false, false));

		shell.setSize(500, 300);
		shell.open();
		shell.pack();
		while (!shell.isDisposed()) {
			if (!display.readAndDispatch()) {
				display.sleep();
			}
		}

		// Dispose of the image and resources
		image.dispose();
		display.dispose();
	}
}
```
image

**Comment by deepika-u** [comment]
> First, this PR is about Linux but you show me Windows screenshort.

To be honest this problem exists on all platforms. Since i have access to only windows platform i have tried to fix the windows problem via this pr.
For Linux, issue is still open => https://github.com/eclipse-platform/eclipse.platform.swt/issues/927
For Mac, issue is open but slightly different, details given in => https://github.com/eclipse-platform/eclipse.platform.swt/issues/928


> Second, please add **improved** test that shows the change under `org.eclipse.swt.tests/ManualTests/org/eclipse/swt/tests/manual`

The snippet to showcase the working behavior is as below. I have added one more case to it when no alignment is set it defaults to CENTER alignment
```
package org.eclipse.swt.tests.manual;
import org.eclipse.swt.SWT;
import org.eclipse.swt.graphics.Image;
import org.eclipse.swt.layout.GridData;
import org.eclipse.swt.layout.GridLayout;
import org.eclipse.swt.widgets.Button;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Shell;

public class Issue0519_ButtonAlignmentExample {

	public static void main(String[] args) {
		Display display = new Display();
		Shell shell = new Shell(display);
		shell.setText("Button Alignment Example for Issue0519");
		shell.setLayout(new GridLayout(1, false));

		// Load an image (you should replace "imagePath" with your image file path)
		//Image image = ImageDescriptor.createFromFile(Snippet.class, "error_tsk.png").createImage();
		Image image = new Image(display, "data/eclipse32.png");

		// Create a button with text and image
		Button button;
		button = new Button(shell, SWT.PUSH);
		button.setText("Right Text");
		button.setImage(image);
		button.setSize(200, 50);
		button.setAlignment(SWT.RIGHT);
		button.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));

		button = new Button(shell, SWT.PUSH);
		button.setText("Left Text");
		button.setImage(image);
		button.setSize(200, 50);
		button.setAlignment(SWT.LEFT);
		button.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));

		button = new Button(shell, SWT.PUSH);
		button.setText("Center Text");
		button.setImage(image);
		button.setSize(200, 50);
		button.setAlignment(SWT.CENTER);

		// Set layout data to take up the whole shell
		button.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));

		button = new Button(shell, SWT.PUSH);
		button.setText("No Alignment assumes CENTER");
		button.setImage(image);
		button.setSize(200, 50);
		//button.setAlignment(SWT.CENTER);
		button.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));

		shell.setSize(500, 300);
		shell.open();
		//shell.pack();
		while (!shell.isDisposed()) {
			if (!display.readAndDispatch()) {
				display.sleep();
			}
		}

		// Dispose of the image and resources
		image.dispose();
		display.dispose();
	}
}
```

I see below behavior for the above snippet
image




> Third, here is the slightly modified example I tried and it shows something unexpected.`

I have tried the same snippet and able to see the difference, can you try it applying again please?
I see as below 
Alignment

Can i go ahead adding this java file to the pr if you dont have any other updations?

@ptziegler : Can you share your observation once again please? Sorry for bothering you again.

**Comment by ptziegler** [comment]
I'm seeing the same as @deepika-u.

This is the example above on this development branch:
image

This is the example on the master:
image

**Comment by deepika-u** [comment]
@iloveeclipse 
Are we good on this pr? can we merge it?

**Comment by iloveeclipse** [comment]
> The snippet to showcase the working behavior is as below. I have added one more case to it when no alignment is set it defaults to CENTER alignment

Once again: can you please add the snippet that demonstrates the problem / fix to the git?

While doing that, please also:
1) Squash all changes to *one* commit.
2) Rebase on *latest* master commit.

**Comment by deepika-u** [comment]
> > The snippet to showcase the working behavior is as below. I have added one more case to it when no alignment is set it defaults to CENTER alignment
> 
> Once again: can you please add the snippet that demonstrates the problem / fix to the git?
> 
> While doing that, please also:
> 
>     1. Squash all changes to _one_ commit.
> 
>     2. Rebase on _latest_ master commit.

@iloveeclipse 
All the suggestions are incorporated, can you please test and confirm?

**Comment by deepika-u** [comment]
I already have Niraj's approval on this pr. Looks like @iloveeclipse is busy. Can any one else review this pr and merge it please? if no other changes are required.

**Review by merks — APPROVED** [review]
It looks like this for me on Windows 10 so I agree this looks good to go:

image

**Comment by merks** [comment]
@iloveeclipse I assume that current implementation, including a test snippet addressed your concerns.   I'll assume all is good and you just don't have time over the course of the day.

**Comment by deepika-u** [comment]
> @iloveeclipse I assume that current implementation, including a test snippet addressed your concerns. I'll assume all is good and you just don't have time over the course of the day.

Yes, as of my understanding everything is already covered. Thanks alot for your approval. Will wait for its merge now :(

**Comment by deepika-u** [comment]
Looks like issue is fixed and working fine.

image

Verified this issue on the below build :
Eclipse SDK
Version: 2024-03 (4.31)
Build id: I20240204-1800
OS: Windows 11, v.10.0, x86_64 / win32
Java vendor: Oracle Corporation
Java runtime version: 21.0.1+12-29
Java version: 21.0.1

**Pull request merged**
